demo/runtime/src/runtime/staking.rs

```diff
@@ -0,0 +1,636 @@
+	impl Externalities for DirectExt {
+		fn get_storage(&self, account: &AccountId, location: &[u8]) -> Option<Vec<u8>> {
+			let mut v = account.to_keyed_vec(STORAGE_OF);
+			v.extend(location);
+			storage::get_raw(&v)
+		}
+		fn get_code(&self, account: &AccountId) -> Vec<u8> {
+			storage::get_raw(&account.to_keyed_vec(CODE_OF)).unwrap_or_default()
+		}
+		fn get_balance(&self, account: &AccountId) -> Balance {
+			storage::get_or_default::<Balance>(&account.to_keyed_vec(BALANCE_OF))
+		}
+	}
+
+	impl<F1, F3, F5> Externalities for Ext<F1, F3, F5> where
+		F1 : Fn(&AccountId, &[u8]) -> Option<Vec<u8>>,
+		F3 : Fn(&AccountId) -> Vec<u8>,
+		F5 : Fn(&AccountId) -> Balance
+	{
+		fn get_storage(&self, account: &AccountId, location: &[u8]) -> Option<Vec<u8>> {
+			(self.do_get_storage)(account, location)
+		}
+		fn get_code(&self, account: &AccountId) -> Vec<u8> {
+			(self.do_get_code)(account)
+		}
+		fn get_balance(&self, account: &AccountId) -> Balance {
+			(self.do_get_balance)(account)
+		}
+	}
+
+	fn commit_state(s: State) {
+		for (address, (maybe_balance, maybe_code, storage)) in s.into_iter() {
+			if let Some(balance) = maybe_balance {
+				storage::put(&address.to_keyed_vec(BALANCE_OF), &balance);
+			}
+			if let Some(code) = maybe_code {
+				storage::put(&address.to_keyed_vec(CODE_OF), &code);
+			}
+			let storage_key = address.to_keyed_vec(STORAGE_OF);
+			for (k, v) in storage.into_iter() {
+				let mut key = storage_key.clone();
+				key.extend(k);
+				if let Some(value) = v {
+					storage::put_raw(&key, &value);
+				} else {
+					storage::kill(&key);
+				}
+			}
+		}
+	}
+
+	fn merge_state(commit_state: State, local: &mut State) {
+		for (address, (maybe_balance, maybe_code, storage)) in commit_state.into_iter() {
+			match local.entry(address) {
+				Entry::Occupied(e) => {
+					let mut value = e.into_mut();
+					if maybe_balance.is_some() {
+						value.0 = maybe_balance;
+					}
+					if maybe_code.is_some() {
+						value.1 = maybe_code;
+					}
+					value.2.extend(storage.into_iter());
+				}
+				Entry::Vacant(e) => {
+					e.insert((maybe_balance, maybe_code, storage));
+				}
+			}
+		}
+	}
+
+	/// Create a smart-contract account.
+	pub fn create(transactor: &AccountId, code: &[u8], value: Balance) {
+		// commit anything that made it this far to storage
+		if let Some(commit) = effect_create(transactor, code, value, DirectExt) {
+			commit_state(commit);
+		}
+	}
+
+	fn effect_create<E: Externalities>(
+		transactor: &AccountId,
+		code: &[u8],
+		value: Balance,
+		ext: E
+	) -> Option<State> {
+		let from_balance = ext.get_balance(transactor);
+		// TODO: a fee.
+		assert!(from_balance >= value);
+
+		let mut dest_pre = blake2_256(code).to_vec();
+		dest_pre.extend(&transactor[..]);
+		let dest = blake2_256(&dest_pre);
+
+		// early-out if degenerate.
+		if &dest == transactor {
+			return None;
+		}
+
+		let mut local = BTreeMap::new();
+
+		// two inserts are safe
+		assert!(&dest != transactor);
```

Comment:
This assert seems redundant and never fails due to early exit above.

Reply:
true, it's there defensively since the two lines following absolutely depend upon that assertion.

Reply:
ah, i see